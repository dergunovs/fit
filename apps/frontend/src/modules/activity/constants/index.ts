import { TActivityChartType } from 'fitness-tracker-contracts';

export const URL_ACTIVITY = '/activities';
export const URL_ACTIVITY_CREATE = `${URL_ACTIVITY}/create`;
export const URL_ACTIVITY_EDIT = `${URL_ACTIVITY}/edit`;

export const URL_ACTIVITY_ADMIN = '/activities_admin';
export const URL_ACTIVITY_ADMIN_EDIT = `${URL_ACTIVITY_ADMIN}/edit`;

export const ACTIVITY_LIST_HEADERS = [{ title: 'Создано' }, { title: 'Пользователь' }];

export const CHART_TYPES: { title: string; value: TActivityChartType }[] = [
  { title: 'Занятия', value: 'activity' },
  { title: 'Подходы', value: 'set' },
  { title: 'Повторы', value: 'repeat' },
  { title: 'Мышцы', value: 'muscle' },
];

export const ACTIVITY_STATISTICS_GAP = 30;
