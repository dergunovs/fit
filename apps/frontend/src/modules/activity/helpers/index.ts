import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
