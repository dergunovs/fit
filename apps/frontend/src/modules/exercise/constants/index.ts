export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;
