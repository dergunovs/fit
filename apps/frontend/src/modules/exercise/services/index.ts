import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
