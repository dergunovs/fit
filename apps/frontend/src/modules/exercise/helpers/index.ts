import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
