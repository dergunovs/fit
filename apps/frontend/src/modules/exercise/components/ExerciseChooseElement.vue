<template>
  <div>
    <UiFlex column gap="16">
      <UiField v-if="props.exercise.isWeights && props.weights?.length" :label="`${t('weight')}, ${t('kg')}`">
        <UiSelect
          v-model="choosenExercise.weight"
          :options="props.weights"
          :lang="locale"
          data-test="exercise-weight"
        />
      </UiField>

      <ExerciseRepeatsChoice
        v-model="choosenExercise.repeats"
        :options="EXERCISE_REPEATS_OPTIONS"
        :title="t('repeat.many')"
        data-test="exercise-repeats"
      />

      <UiFlex justify="space-between">
        <UiButton @click="addExercise(1)" data-test="exercise-add-1">{{ t('add') }}</UiButton>

        <UiFlex>
          <UiButton @click="addExercise(2)" layout="secondary" isNarrow data-test="exercise-add-2">+2</UiButton>
          <UiButton @click="addExercise(3)" layout="secondary" isNarrow data-test="exercise-add-3">+3</UiButton>
        </UiFlex>
      </UiFlex>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { useI18n } from 'vue-i18n';
import { IExercise, IExerciseChoosen, IUser } from 'fitness-tracker-contracts';
import { UiButton, UiField, UiFlex, UiSelect } from 'mhz-ui';
import { createTempId } from 'mhz-helpers';

import ExerciseRepeatsChoice from '@/exercise/components/ExerciseRepeatsChoice.vue';

import { EXERCISE_REPEATS_DEFAULT, EXERCISE_REPEATS_OPTIONS } from '@/exercise/constants';
import { getDefaultExerciseWeight } from '@/exercise/helpers';

interface IProps {
  exercise: IExercise;
  user: IUser;
  weights?: number[];
}

interface IEmit {
  add: [choosenExercise: IExerciseChoosen];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useI18n();

const choosenExercise = ref({
  repeats: EXERCISE_REPEATS_DEFAULT,
  weight: getDefaultExerciseWeight(props.exercise, props.user, props.weights),
});

function addExercise(count: number) {
  for (let i = 0; i < count; i++) {
    const exercise = {
      _id: props.exercise?._id,
      title: props.exercise?.title,
      title_en: props.exercise?.title_en,
      isWeights: props.exercise?.isWeights,
      isWeightsRequired: props.exercise?.isWeightsRequired,
    };

    emit('add', { _id: createTempId(), exercise, ...choosenExercise.value });
  }
}
</script>
