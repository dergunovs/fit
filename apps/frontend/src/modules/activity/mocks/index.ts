import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
