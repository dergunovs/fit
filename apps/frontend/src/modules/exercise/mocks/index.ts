import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
