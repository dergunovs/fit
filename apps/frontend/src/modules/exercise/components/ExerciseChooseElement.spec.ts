import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import ExerciseChooseElement from './ExerciseChooseElement.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISE_FIXTURE } from '@/exercise/fixtures';
import { EXERCISE_REPEATS_DEFAULT, EXERCISE_REPEATS_OPTIONS } from '@/exercise/constants';
import { mockTempId } from '@/common/mocks';
import { USER_FIXTURE, USER_FIXTURE_2 } from '@/user/fixtures';
import { getAvailableExerciseWeights, getDefaultExerciseWeight } from '@/exercise/helpers';

const exerciseRepeats = dataTest('exercise-choosen-repeats');
const exerciseWeight = dataTest('exercise-choosen-weight');
const exerciseAdd1 = dataTest('exercise-add-1');
const exerciseAdd2 = dataTest('exercise-add-2');
const exerciseAdd3 = dataTest('exercise-add-3');

const weights = getAvailableExerciseWeights(EXERCISE_FIXTURE, USER_FIXTURE);
const defaultWeight = getDefaultExerciseWeight(EXERCISE_FIXTURE, USER_FIXTURE, weights);

let wrapper: VueWrapper<InstanceType<typeof ExerciseChooseElement>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseChooseElement, { exercise: EXERCISE_FIXTURE, user: USER_FIXTURE });
});

enableAutoUnmount(afterEach);

describe('ExerciseChooseElement', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseChooseElement)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('sets default repeat and repeats options', async () => {
    expect(wrapper.find(exerciseRepeats).attributes('options')).toBe(EXERCISE_REPEATS_OPTIONS.join(',').toString());
    expect(wrapper.find(exerciseRepeats).attributes('modelvalue')).toBe(EXERCISE_REPEATS_DEFAULT.toString());
  });

  it('sets default weight and weight options', async () => {
    expect(wrapper.find(exerciseWeight).attributes('modelvalue')).toBe(defaultWeight.toString());
  });

  it('emits choosen exercise', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('add');

    const choosenExercise = {
      _id: mockTempId,
      exercise: {
        _id: EXERCISE_FIXTURE._id,
        title: EXERCISE_FIXTURE.title,
        title_en: EXERCISE_FIXTURE.title_en,
        isWeights: EXERCISE_FIXTURE.isWeights,
        isWeightsRequired: EXERCISE_FIXTURE.isWeightsRequired,
        equipmentForWeight: EXERCISE_FIXTURE.equipmentForWeight,
      },
      repeats: EXERCISE_REPEATS_DEFAULT,
      weight: defaultWeight,
    };

    await wrapper.find(exerciseAdd1).trigger('click');

    expect(wrapper.emitted('add')).toHaveLength(1);
    expect(wrapper.emitted()['add'][0]).toStrictEqual([[choosenExercise]]);

    await wrapper.find(exerciseAdd2).trigger('click');

    expect(wrapper.emitted('add')).toHaveLength(2);
    expect(wrapper.emitted()['add'][1]).toStrictEqual([[choosenExercise, choosenExercise]]);

    await wrapper.find(exerciseAdd3).trigger('click');

    expect(wrapper.emitted('add')).toHaveLength(3);
    expect(wrapper.emitted()['add'][2]).toStrictEqual([[choosenExercise, choosenExercise, choosenExercise]]);
  });

  it('does not render weight field when exercise does not require weights', async () => {
    const exerciseWithoutWeights = { ...EXERCISE_FIXTURE, isWeights: false };

    await wrapper.setProps({ exercise: exerciseWithoutWeights, user: USER_FIXTURE });

    expect(wrapper.find(exerciseWeight).exists()).toBe(false);
  });

  it('does not render weight field when no weights available', async () => {
    await wrapper.setProps({ exercise: EXERCISE_FIXTURE, user: USER_FIXTURE_2 });

    expect(wrapper.find(exerciseWeight).exists()).toBe(false);
  });
});
