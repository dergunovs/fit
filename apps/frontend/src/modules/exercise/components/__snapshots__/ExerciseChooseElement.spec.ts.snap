// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseChooseElement > matches snapshot 1`] = `
"<div>
  <div column="" gap="16">
    <fieldset label="Вес, кг"><input type="text" modelvalue="1" options="1,2,3" lang="ru" data-test="exercise-weight"></fieldset>
    <exercise-repeats-choice-stub modelvalue="12" options="8,10,12,15,20" title="Повторы" istall="false" data-test="exercise-repeats"></exercise-repeats-choice-stub>
    <div justify="space-between"><button data-test="exercise-add-1">Добавить</button>
      <div><button layout="secondary" isnarrow="" data-test="exercise-add-2">+2</button><button layout="secondary" isnarrow="" data-test="exercise-add-3">+3</button></div>
    </div>
  </div>
</div>"
`;
